File: src/main/java/com/jloom/compose/RecipeComposer.java
package com.jloom.compose;

import com.jloom.registry.CatalogRoots;
import com.jloom.registry.ModuleManifest;
import com.jloom.util.Tokens;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.LoaderOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.SafeConstructor;
import org.yaml.snakeyaml.nodes.Tag;
import org.yaml.snakeyaml.representer.Representer;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class RecipeComposer {

    private static final String RECIPE_NAME = "com.jloom.generated.ApplyModules";

    public String compose(List<ModuleSelection> modules) {
        return composeFragments(modules.stream()
                .map(s -> new Fragment(s.manifest().id(), s.manifest().mergeRecipes(), s.answers()))
                .toList());
    }

    public String composeUpgrade(List<UpgradeStep> steps) {
        return composeFragments(steps.stream()
                .map(s -> new Fragment(s.moduleId(), List.of(s.recipeResourcePath()), s.answers()))
                .toList());
    }

    public String recipeName() {
        return RECIPE_NAME;
    }

    private String composeFragments(List<Fragment> fragments) {
        List<Map<String, Object>> recipeList = new ArrayList<>();

        for (Fragment fragment : fragments) {
            for (String resourcePath : fragment.resourcePaths()) {
                String text = readFragment(fragment.moduleId(), resourcePath);
                String substituted = Tokens.substitute(text, fragment.answers());
                Object parsed = YAML.load(substituted);
                if (!(parsed instanceof List<?> list)) {
                    throw new IllegalStateException("Fragment " + fragment.moduleId() + "/"
                            + resourcePath + " is not a YAML list (got " + (parsed == null ? "null"
                            : parsed.getClass().getSimpleName()) + ")");
                }
                for (Object entry : list) {
                    if (!(entry instanceof Map<?, ?> entryMap)) {
                        throw new IllegalStateException("Fragment " + fragment.moduleId() + "/"
                                + resourcePath + " contains a non-map entry: " + entry);
                    }
                    @SuppressWarnings("unchecked")
                    Map<String, Object> recipe = (Map<String, Object>) entryMap;
                    recipeList.add(recipe);
                }
            }
        }

        DumperOptions options = new DumperOptions();
        options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
        options.setIndentWithIndicator(true);
        options.setIndicatorIndent(2);
        options.setIndent(2);
        Representer representer = new Representer(options);
        representer.addClassTag(java.util.Map.class, Tag.MAP);
        representer.addClassTag(java.util.List.class, Tag.SEQ);
        representer.addClassTag(String.class, Tag.STR);
        Yaml emitter = new Yaml(representer, options);

        Map<String, Object> document = new LinkedHashMap<>();
        document.put("type", "specs.openrewrite.org/v1beta/recipe");
        document.put("name", RECIPE_NAME);
        document.put("displayName", "jloom generated composite recipe");
        document.put("recipeList", recipeList);
        return emitter.dump(document);
    }

    private static final Yaml YAML = new Yaml(new SafeConstructor(new LoaderOptions()));

    private String readFragment(String moduleId, String resourcePath) {
        try (InputStream in = CatalogRoots.open(moduleId, resourcePath)) {
            if (in == null) {
                throw new IllegalStateException("Module '" + moduleId + "' declares recipe '"
                        + resourcePath + "' but no such resource exists");
            }
            return new String(in.readAllBytes(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private record Fragment(String moduleId, List<String> resourcePaths, Map<String, String> answers) {
    }

    public record ModuleSelection(ModuleManifest manifest, Map<String, String> answers) {
    }

    public record UpgradeStep(String moduleId, String recipeResourcePath, Map<String, String> answers) {
    }
}

File: node/src/merge/recipeComposer.ts
import yaml from "js-yaml";
import { readText } from "../registry/catalogRoots.ts";
import { substitute } from "./tokens.ts";
import type { ModuleManifest } from "../registry/types.ts";
import type { MergeOperation } from "./types.ts";

export interface ModuleSelection {
  manifest: ModuleManifest;
  answers: Record<string, string>;
}

export interface UpgradeStep {
  moduleId: string;
  recipeResourcePath: string;
  answers: Record<string, string>;
}

interface Fragment {
  moduleId: string;
  resourcePaths: string[];
  answers: Record<string, string>;
}

/** Mirrors RecipeComposer.compose(...) / composeUpgrade(...) — returns the flat operation
 * list directly rather than re-serializing to a YAML document, since nothing external (no
 * Gradle/OpenRewrite subprocess) needs to read it back as text anymore. */
export function compose(modules: ModuleSelection[]): MergeOperation[] {
  return composeFragments(
    modules.map((s) => ({
      moduleId: s.manifest.id,
      resourcePaths: s.manifest.mergeRecipes,
      answers: s.answers,
    })),
  );
}

export function composeUpgrade(steps: UpgradeStep[]): MergeOperation[] {
  return composeFragments(
    steps.map((s) => ({
      moduleId: s.moduleId,
      resourcePaths: [s.recipeResourcePath],
      answers: s.answers,
    })),
  );
}

function composeFragments(fragments: Fragment[]): MergeOperation[] {
  const operations: MergeOperation[] = [];

  for (const fragment of fragments) {
    for (const resourcePath of fragment.resourcePaths) {
      const text = readFragment(fragment.moduleId, resourcePath);
      const substituted = substitute(text, fragment.answers);
      const parsed = yaml.load(substituted);
      if (!Array.isArray(parsed)) {
        throw new Error(
          `Fragment ${fragment.moduleId}/${resourcePath} is not a YAML list (got ${typeof parsed})`,
        );
      }
      for (const entry of parsed) {
        operations.push(parseOperation(fragment.moduleId, resourcePath, entry));
      }
    }
  }

  return operations;
}

function readFragment(moduleId: string, resourcePath: string): string {
  const text = readText(moduleId, resourcePath);
  if (text === undefined) {
    throw new Error(`Module '${moduleId}' declares recipe '${resourcePath}' but no such resource exists`);
  }
  return text;
}

function parseOperation(moduleId: string, resourcePath: string, entry: unknown): MergeOperation {
  if (entry === null || typeof entry !== "object") {
    throw new Error(`Fragment ${moduleId}/${resourcePath} contains a non-map entry: ${String(entry)}`);
  }
  const keys = Object.keys(entry as Record<string, unknown>);
  if (keys.length !== 1) {
    throw new Error(`Fragment ${moduleId}/${resourcePath} contains a malformed recipe entry: ${JSON.stringify(entry)}`);
  }
  const recipeType = keys[0]!;
  const params = (entry as Record<string, unknown>)[recipeType] as Record<string, unknown>;

  switch (recipeType) {
    case "org.openrewrite.gradle.AddDependency":
      return {
        type: recipeType,
        groupId: String(params.groupId),
        artifactId: String(params.artifactId),
        version: params.version === undefined ? undefined : String(params.version),
        configuration: String(params.configuration),
      };
    case "org.openrewrite.yaml.MergeYaml":
      return {
        type: recipeType,
        key: String(params.key),
        yaml: String(params.yaml),
        filePattern: String(params.filePattern),
      };
    case "org.openrewrite.yaml.ChangePropertyKey":
      return {
        type: recipeType,
        oldPropertyKey: String(params.oldPropertyKey),
        newPropertyKey: String(params.newPropertyKey),
        filePattern: String(params.filePattern),
      };
    case "org.openrewrite.text.CreateTextFile":
      return {
        type: recipeType,
        relativeFileName: String(params.relativeFileName),
        fileContents: String(params.fileContents),
        overwriteExisting: params.overwriteExisting === true,
      };
    default:
      throw new Error(`Fragment ${moduleId}/${resourcePath} uses unsupported recipe type: ${recipeType}`);
  }
}
